### ISSUE TITLE
Mac/osx: APY_ANKI_PATH related import error

### ISSUE THREAD

**Issue by Shikeiku** [lervag/apy#24]
Hi,

My goal was to install `Anki-2.1.26` and `apy-0.6.0` on my system python (`brew install python@3.8`), and saving a venv to at least know the python requirements/dependencies that got everything working.

I downloaded the Anki-source code, instead of cloning the git repo and checking out the 2.1.26 version. Which looked like this.

```shell
┏╸~/bin/anki-2.1.26
┗╸❯❯❯ ls
.                   README.contributing pylib
..                  README.development  qt
.gitattributes      README.md           rslib
.github             dist                rspy
.gitignore          meta                run
CONTRIBUTORS        pkgkey.asc          scripts
LICENSE             proto
Makefile            pyenv
```

Note that `~/bin` is on my path. After installing all dependencies and sorting out errors from running the `~/bin/anki-2.1.26/run` executable, my `pip3 list` looks similar to

```shell
┏╸~/bin/anki-2.1.26
┗╸❯❯❯ pip3 list
Package        Version
-------------- ---------
ankirspy       2.1.26
appdirs        1.4.4
attrs          20.1.0
beautifulsoup4 4.9.1
certifi        2020.6.20
chardet        3.0.4
click          7.1.2
decorator      4.4.2
distlib        0.3.1
filelock       3.0.12
idna           2.10
jsonschema     3.2.0
Markdown       3.2.2
pip            20.1.1
protobuf       3.13.0
PyQt5          5.15.0
PyQt5-sip      12.8.0
PyQtWebEngine  5.15.0
pyrsistent     0.16.0
readchar       2.0.1
requests       2.24.0
Send2Trash     1.5.0
setuptools     49.2.0
six            1.15.0
soupsieve      2.0.1
urllib3        1.25.10
virtualenv     20.0.31
wheel          0.34.2
```

Next I cloned and installed `apy-0.6.0`, by using `pip3 install .` inside the apy directory, adding it to the list above. I also configured the apy paths to

```shell
┏╸~/bin/anki-2.1.26
┗╸❯❯❯ env | grep APY
APY_CONFIG=/Users/mikevink/.dotfiles/apy/apy.json
APY_ANKI_PATH=/Users/mikevink/bin/anki-2.1.26

┏╸~/bin/anki-2.1.26
┗╸❯❯❯ cat ~/.dotfiles/apy/apy.json
{
  "base": "/Users/mikevink/Library/Application Support/Anki2"
}
```

Launching Anki with `~/bin/anki-2.1.26/run` works fine at this point, but note that `which anki` gives anki not found, and when I run apy I get
```shell
┏╸~/bin/anki-2.1.26
┗╸❯❯❯ apy
Traceback (most recent call last):
  File "/usr/local/bin/apy", line 5, in <module>
    from apy.cli import main
  File "/usr/local/lib/python3.8/site-packages/apy/cli.py", line 8, in <module>
    from apy.anki import Anki
  File "/usr/local/lib/python3.8/site-packages/apy/anki.py", line 8, in <module>
    import anki
ModuleNotFoundError: No module named 'anki'
```

My _"solution"_ so far is to set `APY_ANKI_PATH=/Users/mikevink/bin/anki-2.1.26/pylib` and `PYTHONPATH=/Users/mikevink/Documents/python:/Users/mikevink/bin/anki-2.1.26/qt`
```shell
┏╸~/bin/anki-2.1.26
┗╸❯❯❯ env | grep anki-2.1.26/
APY_ANKI_PATH=/Users/mikevink/bin/anki-2.1.26/pylib
PYTHONPATH=/Users/mikevink/Documents/python:/Users/mikevink/bin/anki-2.1.26/qt
```

And now apy and Anki works on my fresh system python! Thanks, I have been enjoying it a lot, I also made this reddit [post](https://www.reddit.com/r/Anki/comments/ickl78/using_vim_and_a_custom_markdown_syntax_to_make/), and if it is oke with you I would like to make a follow up post showing apy in action! 

Hope I didn't do something stupid since I'm a little bit of a noob, in any case this shows that currently everything works on Mac/osx, maybe I could make a pull request for building Anki and apy on Mac.

**Comment by lervag** [comment]
> Hi,

Hi, nice to meet you here as well!

> My goal was to install `Anki-2.1.26` and `apy-0.6.0` on my system python (`brew install python@3.8`), and saving a venv to at least know the python requirements/dependencies that got everything working.

Seems like a sensible goal!

> I downloaded the Anki-source code, instead of …

Thanks for showing and explaining how you made things work on macOS! I took the liberty of making some minor changes to your post to clarify that you are working within a virtualenv. Please feel free to re-edit if I was wrong!

> Launching Anki with `~/bin/anki-2.1.26/run` works fine at this point, but note that `which anki` gives anki not found, and when I run apy I get

I assume this is because there is no `anki` executable in the repo, and that this executable is created by the install script.

> My _"solution"_ so far is to set `APY_ANKI_PATH=/Users/mikevink/bin/anki-2.1.26/pylib` and `PYTHONPATH=/Users/mikevink/Documents/python:/Users/mikevink/bin/anki-2.1.26/qt`

Yes, this looks sensible. First, `APY_ANKI_PATH` must point to the anki Python package, and the Anki qt stuff must be available as well.

> And now apy and Anki works on my fresh system python! Thanks, I have been enjoying it a lot, I also made this reddit [post](https://www.reddit.com/r/Anki/comments/ickl78/using_vim_and_a_custom_markdown_syntax_to_make/), and if it is oke with you I would like to make a follow up post showing apy in action!

> Hope I didn't do something stupid since I'm a little bit of a noob, in any case this shows that currently everything works on Mac/osx, maybe I could make a pull request for building Anki and apy on Mac.

Although I do now have quite a bit of experience with Python, I still have to admit that Python packaging and dependencies can be confusing. At first, I only made `apy` with myself in mind. But as it started to work well I thought it would be nice to share it. Then I realized I needed to make things more robust, and someone suggested that I started to apply proper versioning. I try to keep things well documented and helpful for others, although I don't want to "dumb it down" too much. That is, I want my time to mostly go towards improvements that are also relevant for myself.

Still, I am quite certain that the packaging stuff can be improved, and at least the docs. So please feel free to suggest changes to e.g. the README file or similar to make things work even more smoothly on e.g. macOS. A PR would be very welcome, but I also accept concrete suggestions that I can commit personally. Thanks!

Finally, I would be more than happy to see a follow up post about `apy` in action. Feel free to ping me if you want pointers on it before submitting.

Do you by chance also use Vim? If so, let me know, and I can share some more of my Vim related `apy` workflow.

**Comment by Shikeiku** [comment]
> But as it started to work well I thought it would be nice to share it. Then I realized I needed to make things more robust, and someone suggested that I started to apply proper versioning.

Thanks for this. :)

>I try to keep things well documented and helpful for others, although I don't want to "dumb it down" too much. That is, I want my time to mostly go towards improvements that are also relevant for myself.

I agree with this, and I consider apy good at what it does after exploring it a bit. Initially I wanted to add stuff to apy, but now I think it is better for users to write their own helper scripts/functions or apy wrappers rather than add things to apy itself and potentially making it worse/ more convoluted.

>Do you by chance also use Vim? If so, let me know, and I can share some more of my Vim related apy workflow.

Yes that would be interesting! :)

>A PR would be very welcome, but I also accept concrete suggestions that I can commit personally. Thanks!

I want to use apy the coming period of my school to get more experience first, then I could create a pr.

Take what follows with a grain of salt since they are just my thoughts after exploring apy for a couple days. :)

## what I discovered after exploring apy for a couple days
### using Cloze model notes with the apy add command automatically adds markdown false in defaults
Mostly I make cloze cards with only two fields, and apply some styling tags in fields, instead of in the model. I first considered proposing configurable _"subfield"_ styling to apy ([Link to my fork's most relevant part](https://github.com/Shikeiku/apy/blob/91f066e9dc5d7963fbb67e5d0cf5fbc5035da030/apy/convert.py#L188-L191
)), where one could give custom tags in the config.json. But now I think that I should just write my own helper scripts to give notes to `apy add-from-file`, instead of changing apy.

For example using my subfield styling tag and `apy add -m Cloze` the question tag adds a html horizontal line, and the answer tag makes the text bigger using markdown span:
```shell
# Note ID: 1598607790727
model: Cloze
deck: Professional::‚ûóMath
tags: marked

## Text
<question>
question
</question>

{{c1::
<answer>My answer</answer>

Some other lines.

\( 1 + 2 = 3\) 

\[
  \int_{\infty}^{- \infty} f(x) \mathop{d x}
  .\]


\```python
print('hello, world')
\```
}}

## context
Book chapter 1 exercise 1
```
After conversion with the markdown_to_html function of my fork, we see the html and span appear that are configured in the `config.json`
```shell
## Text
question
<hr style='height:2px;border-width:0;color:gray;background-color:gray'>

{{c1::
<span style='font-size:30px'>My answer</span>

Some other lines.

\( 1 + 2 = 3\)

\[
  \int_{\infty}^{- \infty} f(x) \mathop{d x}
  .\]


\```python
print('hello, world')
\```
}}

## context
Book chapter 1 exercise 1
```

Final result in Anki, where everything below the rule is a cloze deletion (note that I haven't changed the background of the pygments code block)
![Screenshot 2020-08-28 at 16 47 13](https://user-images.githubusercontent.com/59492084/91580992-2f639e00-e94e-11ea-9495-10b15a4a602d.png)

As a final remark to that, is that when using `apy add -m Cloze` for example the card is tagged as `markdown: false` in defaults. which I changed in my fork since I wanted to play around with making **markdown Cloze** cards using the add command.
https://github.com/lervag/apy/blob/d5f9257995b45ed79377d500a9ef9c73931dd5d6/apy/anki.py#L356-L357

### sidenote: when n_decks is greater than 10, `apy add -d ask` doesn't work for all decks
Also want to mention that I have over ten decks and using `apy add -d ask` doesn't let me choose a deck option that is greater than 10
```shell
$ apy add -d ask
reverting to stock json
Choose deck:
1: Chess
2: Chess::Drills
3: Chess::My Games
4: Chess::Tactics
5: Default1
6: Hobby
7: Hobby::russian living language course
8: Hobby::🇬🇧English
9: Hobby::🇷🇺Russian
10: Hobby::📚Books
11: Hobby::🤔Logic, Fallacies, and biases
12: Professional
13: Professional::➗Math
14: Professional::🏹Math exercises
15: Professional::🧪Chemistry
16: inbox anki-cloze
17: inbox anki-img
> 1
```

**Comment by lervag** [comment]
> > But as it started to work well I thought it would be nice to share it. Then I realized I needed to make things more robust, and someone suggested that I started to apply proper versioning.
> 
> Thanks for this. :)

Your welcome :)

> > I took the liberty of making some minor changes to your post to clarify that you are working within a virtualenv. Please feel free to re-edit if I was wrong!
> 
> I was wrong about the virtual env, I'm just using everything in my system python. I was confused myself about that, probably can remove mentioning virtualenvs entirely.

Ah, ok. Yes, I think it could be helpful to others if you update your post to reflect an actual, working installation process.

> > Do you by chance also use Vim? If so, let me know, and I can share some more of my Vim related apy workflow.
> 
> Yes that would be interesting! :)

Ok! There were several different threads in your post, and I have a lot of stuff going on. Could I bother you to open separate issues for the various things? E.g. one issue about Vim workflow, it would only need to be a short question, and I'll resolve it by outlining my workflow. Then I'll be able to reply in detail when I get the time.

> > A PR would be very welcome, but I also accept concrete suggestions that I can commit personally. Thanks!
> 
> I want to use apy the coming period of my school to get more experience first, then I could create a pr. I guess this issue doesn't need to stay open also.

Great, sounds good!

> ### using Cloze model notes with the apy add command automatically adds markdown false in defaults

I am sorry to say I did not quite understand this part. I get the feeling you are describing a workflow, and that there is something with `apy` that works not quite as you expect/want? If I understand correctly, I also think my Vim workflow would be interesting for you, as it involves turning simple things like

```
Q: question
A: answer
Q: question2
A: answer2
```

into

```markdown
# Note 1
opts

## Front
**heading**

question

## Back
answer

# Note 2
...
```

You also mention that the `markdown: false` is added as a default setting for the Cloze type (indeed, for anything not `Basic`). This is probably surprising, since it is based on my personal workflow.

For the `add-single` command, one can specify configuration presets and also modify the default configuration. This should probably be expanded to also work for the other `add` variants. I think that could make the default `markdown: false` line fully unnecessary. Feel free to open an issue about this if you think it is useful!

> ... (note that I haven't changed the background of the pygments code block)

I personally find it useful to quickly edit the CSS from apy either through the `apy review` feature or with the `apy model edit-css`. It does not give immediate feedback though, so for e.g. specifying colors it is still useful to open Anki and modify the CSS scheme in the card layout screens.

> Also want to mention that I have over ten decks and using `apy add -d ask` doesn't let me choose a deck option that is greater than 9

Ah, sorry! I have a single deck for everything, and didn't really consider people having a lot of different decks. Could you open a new issue for this?

**Comment by lervag** [comment]
I'll close this, but feel free to continue discussion. Also feel free to open new issues/feature requests with more concrete suggestions/ideas, as mentioned in my previous post.

**Comment by Shikeiku** [comment]
Thanks for the replies, there were indeed a lot of threads in my post. The two main issues that I could do a better job specifying are the configuration preset for the add command and having more than ten decks. This was my first github issue. :)

**Comment by lervag** [comment]
> This was my first github issue. :)

Then I'd say it was a good start! :)
